### DISCUSSION TITLE
Updating the pinmux for the Xavier NX

### DISCUSSION THREAD

**Discussion by octonout** [OE4T/meta-tegra#989] [Q&A]
Hi. I am trying to build an image to boot the Xavier NX on our custom carrier board. I have modified the pinmux according to the changes on our carrier board. But I haven't figured out what I am supposed to do with the .dtsi files and the .cfg files that are created in the process of Platform Adaptation. For the Jetson Nano I used to follow the following steps:

1. Clone https://github.com/OE4T/linux-tegra-4.9/tree/patches-l4t-r32.4
2. Replace the existing pinmux and gpio .dtsi with excel sheet generated pinmux and gpio .dtsi files in `nvidia/platform/t210/porg/kernel-dts/porg-platforms`
3. Make changes to other .dtsi files where needed
4. Create a patch for the changes
5. Apply the patch to the kernel using linux-tegra_4.9.bbappend recipe

For the Xavier NX there I had to create .cfg files from the .dtsi files using pinmux-dts2cfg.py. From what I have read on the forums, these files are used to apply the pinmux by the bootloader. And that the kernel itself does not use these. Also there are no default pinmux and gpio dtsi files in `nvidia/platform/t19x/jakku/kernel-dts` like there are in `nvidia/platform/t210/porg/kernel-dts/porg-platforms`.

I did trawl through the meta-tegra sources to find where the .cfg file might have been used. `sources/meta-tegra/conf/machine/include/xavier-nx.inc` has the following:
```
EMMC_BCT ?= "tegra194-mb1-bct-memcfg-p3668-0001-a00.cfg"
EMMC_BCT_OVERRIDE ?= "tegra194-memcfg-sw-override.cfg"
```
Also, it seems like the BSP sources are downloaded using `sources/meta-tegra/classes/l4t_bsp.bbclass`. Not very sure whether the recipe `sources/meta-tegra/recipes-bsp/trusty/trusty-l4t_32.4.3.bb` is responsible for downloading the BSP.

I was wondering whether I am supposed create a .bbappend for the `trusty-l4t_32.4.3.bb` to put the .cfg files I created into BSP sources once they are created. Also whether I have to name the files in a certain way or replace one of the existing files with my own .cfg files.

The file `sources/meta-tegra/recipes-bsp/tegra-binaries/tegra-flashvars/xavier-nx/flashvars` has the following variables:
```
FLASHVARS="BPFDTB_FILE PMIC_CONFIG SCR_COLD_BOOT_CONFIG MISC_COLD_BOOT_CONFIG PINMUX_CONFIG PMC_CONFIG UPHY_CONFIG DEVICE_CONFIG GPIOINT_CONFIG PROD_CONFIG SCR_CONFIG BR_CMD_CONFIG DEV_PARAMS"
BPFDTB_FILE="tegra194-a02-bpmp-p3668-a00.dtb"
PMIC_CONFIG="tegra194-mb1-bct-pmic-p3668-0001-a00.cfg"
SCR_COLD_BOOT_CONFIG="tegra194-mb1-bct-scr-cbb-mini-p3668.cfg"
MISC_COLD_BOOT_CONFIG="tegra194-mb1-bct-misc-l4t.cfg"
PINMUX_CONFIG="tegra19x-mb1-pinmux-p3668-a01.cfg"
PMC_CONFIG="tegra19x-mb1-padvoltage-p3668-a01.cfg"
UPHY_CONFIG=""
DEVICE_CONFIG="tegra19x-mb1-bct-device-qspi-p3668.cfg"
GPIOINT_CONFIG="tegra194-mb1-bct-gpioint-p3668-0001-a00.cfg"
PROD_CONFIG="tegra19x-mb1-prod-p3668-0001-a00.cfg"
SCR_CONFIG="tegra194-mb1-bct-scr-cbb-mini-p3668.cfg"
BR_CMD_CONFIG="tegra194-mb1-bct-reset-p3668-0001-a00.cfg"
DEV_PARAMS="tegra194-br-bct-qspi.cfg"
```

I am wondering whether these have to be edited if I have custom names for my .cfg files.

Would really appreciate if someone can tell me what I need to do with these .dtsi and .cfg files.

Thanks.

**Comment by madisongh** [comment]
The pinmux and GPIO configurations are applied by the early-stage bootloaders on the TX2 and Xavier platforms (not Trusty).  There's a brief description of what to do on [this wiki page](https://github.com/OE4T/meta-tegra/wiki/Creating-a-custom-MACHINE).  It's not the `trusty` recipe you modify, but `tegra-flashvars`, to provide your customized `flashvars` file that points to the `.cfg` files you generate with the `pinmux-dts2cfg.py` tool. You also need a recipe to make those `.cfg` files available, and a bbappend for the `tegra-bootfiles` recipe to install those `.cfg` files into `${datadir}/tegraflash/${MACHINE}`.
